lxd/devlxd_storage.go

```diff
@@ -0,0 +1,513 @@
+	"github.com/canonical/lxd/shared/entity"
+)
+
+// devLXDStoragePoolVolumeTypeAccessHandler returns an access handler which checks the given entitlement
+// on a storage volume and ensures cross-project access is not allowed.
+func devLXDStoragePoolVolumeTypeAccessHandler(entityType entity.Type, entitlement auth.Entitlement) func(d *Daemon, r *http.Request) response.Response {
+	return func(d *Daemon, r *http.Request) response.Response {
+		s := d.State()
+
+		// Disallow cross-project access and ensure project query parameter is set.
+		err := enforceDevLXDProject(r)
+		if err != nil {
+			return response.DevLXDErrorResponse(err)
+		}
+
+		err = checkStoragePoolVolumeTypeAccess(s, r, entityType, entitlement)
+		if err != nil {
+			return response.DevLXDErrorResponse(err)
+		}
+
+		return response.EmptySyncResponse
+	}
+}
+
+var devLXDStoragePoolEndpoint = APIEndpoint{
+	Path: "storage-pools/{poolName}",
+	Get:  APIEndpointAction{Handler: devLXDStoragePoolGetHandler, AccessHandler: allowDevLXDAuthenticated},
+}
+
+// devLXDStoragePoolGetHandler retrieves information about the specified storage pool.
+func devLXDStoragePoolGetHandler(d *Daemon, r *http.Request) response.Response {
+	inst, err := getInstanceFromContextAndCheckSecurityFlags(r.Context(), devLXDSecurityKey, devLXDSecurityManagementVolumesKey)
+	if err != nil {
+		return response.DevLXDErrorResponse(err)
+	}
+
+	poolName, err := url.PathUnescape(mux.Vars(r)["poolName"])
+	if err != nil {
+		return response.DevLXDErrorResponse(api.NewGenericStatusError(http.StatusBadRequest))
+	}
+
+	// Get storage pool.
+	projectName := inst.Project().Name
+	pool := api.StoragePool{}
+
+	url := api.NewURL().Path("1.0", "storage-pools", poolName).Project(projectName)
+	req, err := lxd.NewRequestWithContext(r.Context(), http.MethodGet, url.String(), nil, "")
+	if err != nil {
+		return response.DevLXDErrorResponse(err)
+	}
+
+	resp := storagePoolGet(d, req)
+	etag, err := response.NewResponseCapture(req).RenderToStruct(resp, &pool)
+	if err != nil {
+		return response.DevLXDErrorResponse(err)
+	}
+
+	// Convert to devLXD response.
+	respPool := api.DevLXDStoragePool{
+		Name:   pool.Name,
+		Driver: pool.Driver,
+		Status: pool.Status,
+	}
+
+	return response.DevLXDResponseETag(http.StatusOK, respPool, "json", etag)
+}
+
+var devLXDStoragePoolVolumesEndpoint = APIEndpoint{
+	Path: "storage-pools/{poolName}/volumes",
+	Get:  APIEndpointAction{Handler: devLXDStoragePoolVolumesGetHandler, AccessHandler: allowDevLXDAuthenticated},
+	Post: APIEndpointAction{Handler: devLXDStoragePoolVolumesPostHandler, AccessHandler: allowDevLXDAuthenticated},
+}
+
+var devLXDStoragePoolVolumesTypeEndpoint = APIEndpoint{
+	Path: "storage-pools/{poolName}/volumes/{type}",
+	Get:  APIEndpointAction{Handler: devLXDStoragePoolVolumesGetHandler, AccessHandler: allowDevLXDAuthenticated},
+	Post: APIEndpointAction{Handler: devLXDStoragePoolVolumesPostHandler, AccessHandler: allowDevLXDAuthenticated},
```

Comment:
This access handler should be checking if the devlxd identity has permission to create volumes in the project.